common/transport/amqp/src/amqp.ts

```diff
@@ -251,44 +413,8 @@ export class Amqp {
    * @param {string}    to        The destination of the message.
    * @param {Function}  done      Called when the message is sent or if an error happened.
    */
-  send(message: Message, endpoint: string, to: string, done: GenericAmqpBaseCallback): void {
-    const self = this;
-    if (!this._connected) {
-      self._safeCallback(done, new errors.NotConnectedError('Cannot send while disconnected.'));
-    } else {
-      /*Codes_SRS_NODE_COMMON_AMQP_16_006: [The `send` method shall construct an AMQP message using information supplied by the caller, as follows:
-      The ‘to’ field of the message should be set to the ‘to’ argument.
-      The ‘body’ of the message should be built using the message argument.] */
-
-      let amqpMessage = AmqpMessage.fromMessage(message);
-      if (to !== undefined) {
-        amqpMessage.properties.to = to;
-      }
-
-      const sendAction = (sender: any, msg: AmqpMessage, done: GenericAmqpBaseCallback): void => {
-        sender.send(msg)
-          .then((state) => {
-            self._safeCallback(done, null, new results.MessageEnqueued(state));
-            return null;
-          })
-          .catch((err) => {
-            /*Codes_SRS_NODE_IOTHUB_AMQPCOMMON_16_007: [If sendEvent encounters an error before it can send the request, it shall invoke the `done` callback function and pass the standard JavaScript Error object with a text description of the error (err.message).]*/
-            self._safeCallback(done, err);
-          });
-      };
-
-      if (!this._senders[endpoint]) {
-        self.attachSenderLink(endpoint, null, (err: Error | null, sender: any): void => {
-          if (err) {
-            self._safeCallback(done, err);
-          } else {
-            sendAction(sender, amqpMessage, done);
-          }
-        });
-      } else {
-        sendAction(this._senders[endpoint], amqpMessage, done);
-      }
-    }
+  send(message: Message, endpoint: string, to: string, done: GenericAmqpBaseCallback<any>): void {
+    this._fsm.handle('send', message, endpoint, to, done);
```

Comment:
>handle [](start = 14, length = 6)

If the user calls send() without ever calling connect(), where does the send error out?  I'm guessing it will be somewhere cryptic that will be difficult for the caller to track down.  Do we want an explicit check to make sure the user called connect() at least once before calling any of the other functions?

Reply:
I hate the fact that we have a send() method here. SenderLink.send() should be used instead (and at some point we'll be able to get rid of the link collections).

---
In reply to: [139533766](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139533766) [](ancestors = 139533766)

Reply:
we can try to connect and send using the FSM but it's likely going to fail if not authenticated.

---
In reply to: [139567812](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139567812) [](ancestors = 139567812,139533766)